File: sdk/typescript-guest/examples/events.ts
import type * as Contract from '../generated/interfaces/tests-nats-events-api.js';
import * as events from '../vendor/lsf/sdk/typescript-guest/capabilities/events.js';
import { unwrap } from '../vendor/lsf/sdk/typescript-guest/capabilities/result.js';
export const api: typeof Contract = {
  run(_which, topic, handle) {
    const result = events.publish({ topic, payload: new Uint8Array([112, 97, 121, 108, 111, 97, 100]),
      mediaType: 'text/plain', attributes: [], idempotencyKey: 'guest-sdk-' + handle.toString() });
    if (result.tag === 'err' && result.val.tag === 'permission-denied') return 10n;
    if (result.tag === 'err' && result.val.tag === 'uncertain') return 11n;
    return unwrap(result).sequence;
  },
};

File: sdk/go-guest/examples/events.go
package export_tests_nats_events_api

import (
	wit "go.bytecodealliance.org/pkg/wit/types"
	"strconv"
	events "wit_component/lsf/events"
)

func Run(_ uint32, topic string, handle uint64) uint64 {
	r := events.Publish(events.Event{Topic: topic, Key: wit.None[string](), Payload: []byte("payload"),
		MediaType: "text/plain", Attributes: []wit.Tuple2[string, string]{},
		IdempotencyKey: "guest-sdk-" + strconv.FormatUint(handle, 10)})
	if r.IsOk() {
		value := r.Ok()
		if value.EventId == "" || value.StreamName == "" {
			panic("incomplete publication receipt")
		}
		return value.Sequence
	}
	switch r.Err().Tag() {
	case events.EventErrorPermissionDenied:
		return 10
	case events.EventErrorUncertain:
		return 11 // preserve uncertainty; never retry
	default:
		panic("unexpected event outcome")
	}
}
